src/main/kotlin/org/codefreak/codefreak/graphql/api/FileApi.kt

```diff
@@ -61,3 +71,52 @@ class FileQuery : BaseResolver(), Query {
     FileDto(answer.id, digest, file)
   }
 }
+
+@Component
+class FileMutation : BaseResolver(), Mutation {
+  fun createFile(fileContext: FileContext, path: String): Boolean = context {
+    authorize(fileContext)
+    serviceAccess.getService(FileService::class).createFile(fileContext.id, path)
+    true
+  }
+
+  fun createDirectory(fileContext: FileContext, path: String): Boolean = context {
+    authorize(fileContext)
+    serviceAccess.getService(FileService::class).createDirectory(fileContext.id, path)
+    true
+  }
+
+  fun uploadFile(fileContext: FileContext, path: String, contents: ApplicationPart): Boolean = context {
+    authorize(fileContext)
+    serviceAccess.getService(FileService::class).filePutContents(fileContext.id, path).use {
+      it.write(contents.inputStream.readBytes())
+    }
+    true
+  }
+
+  fun moveFile(fileContext: FileContext, sourcePath: String, targetPath: String): Boolean = context {
+    authorize(fileContext)
+    serviceAccess.getService(FileService::class).moveFile(fileContext.id, sourcePath, targetPath)
+    true
+  }
+
+  fun deleteFile(fileContext: FileContext, path: String): Boolean = context {
+    authorize(fileContext)
+    serviceAccess.getService(FileService::class).deleteFile(fileContext.id, path)
+    true
+  }
+
+  private fun authorize(fileContext: FileContext) = context {
+    when (fileContext.contextType) {
+      FileContextType.ANSWER -> {
+        val answer = serviceAccess.getService(AnswerService::class).findAnswer(fileContext.id)
+        authorization.requireAuthorityIfNotCurrentUser(answer.task.owner, Authority.ROLE_ADMIN)
+      }
+      FileContextType.TASK -> {
+        val task = serviceAccess.getService(TaskService::class).findTask(fileContext.id)
+        authorization.requireAuthorityIfNotCurrentUser(task.owner, Authority.ROLE_ADMIN)
+      }
+    }
+    true
```

Comment:
Auth must fail for missing context type

Reply:
`FileContextType` is an enumeration and not optional, can it be missing?

Reply:
What I mean is: if we add another entry to the enum and forget to edit this method, auth will always pass for the new type. This is because the `when` is not used as an expression and had no `else` branch.
See https://proandroiddev.com/til-when-is-when-exhaustive-31d69f630a8b
I think we should use an `exhaustive` extension like shown in the article to get a compiler error.

Reply:
Oh, okay.